File: src/predict.R
# src/predict.R
library(raster)
library(randomForest)

predict_landcover <- function(model_path, input_tif, output_png) {
  rf_model <- readRDS(model_path)
  raster_data <- stack(input_tif)
  ndvi <- (raster_data[[4]] - raster_data[[1]]) / (raster_data[[4]] + raster_data[[1]])
  X_new <- data.frame(nir = raster_data[[4]][], red = raster_data[[1]][])
  
  predictions <- predict(rf_model, X_new)
  pred_matrix <- matrix(predictions, nrow = nrow(raster_data), ncol = ncol(raster_data))
  
  ggplot() +
    geom_raster(data = as.data.frame(rasterToPoints(pred_matrix)), aes(x = x, y = y, fill = layer)) +
    scale_fill_manual(values = rainbow(length(unique(predictions)))) +
    ggtitle("Land Cover Classification")
  ggsave(output_png)
}

predict_landcover("models/rf_model.rds", "data/landcover.tif", "outputs/classified_map.png")

File: src/predict.py
# src/predict.py
import rasterio
import numpy as np
import joblib
import matplotlib.pyplot as plt

def predict_landcover(model_path, input_tif, output_png):
    clf = joblib.load(model_path)
    
    with rasterio.open(input_tif) as src:
        nir = src.read(4)
        red = src.read(1)
    
    ndvi = (nir.astype(float) - red.astype(float)) / (nir + red)
    X_new = np.column_stack((nir.ravel(), red.ravel()))
    
    predictions = clf.predict(X_new).reshape(nir.shape)
    
    plt.imshow(predictions, cmap='tab10')
    plt.colorbar(label='Land Cover Class')
    plt.title("Land Cover Classification")
    plt.savefig(output_png)
    plt.show()

predict_landcover("models/rf_model.pkl", "data/landcover.tif", "outputs/classified_map.png")